scripts/irods/test/test_prep_genquery_iterator.py

```diff
@@ -832,3 +832,155 @@ def tearDown(self):
+                rows = 0
+                for r in Query(callback, 'DATA_NAME', "COLL_NAME = '{self.admin.session_collection}' and DATA_NAME like '{data_name_prefix}.%'", offset=3, parser=Parser.GENQUERY2):
+                    rows += 1
+                callback.writeLine('stdout', f'count: {{rows}}')
+            INPUT null
+            OUTPUT ruleExecOut
+            '''))
+
+        # Execute the rule.
+        rep_instance = IrodsConfig().default_rule_engine_plugin + '-instance'
+        self.admin.assert_icommand(['irule', '-r', rep_instance, '-F', rule_file], 'STDOUT', ['count: 7'])
+
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-python', 'Requires PREP.')
+    def test_genquery_iterator_supports_genquery2_limit__issue_7909(self):
+        # Create some data objects.
+        data_name_prefix = 'test_genquery_iterator_supports_genquery2_limit__issue_7909'
+        for i in range(10):
+            self.admin.assert_icommand(['itouch', f'{data_name_prefix}.{i}'])
+
+        # Create a rule file which, when executed, lists the data objects in the admin's
+        # session collection, but limits the resultset to only two entries.
+        rule_file = f'{self.admin.local_session_dir}/test_genquery_iterator_supports_genquery2_limit__issue_7909.r'
+        with open(rule_file, 'w') as rf:
+            rf.write(dedent(f'''
+            def main(rule_args, callback, rei):
+                from genquery import Query, Parser
+                rows = 0
+                for r in Query(callback, 'DATA_NAME', "COLL_NAME = '{self.admin.session_collection}' and DATA_NAME like '{data_name_prefix}.%'", limit=2, parser=Parser.GENQUERY2):
+                    rows += 1
+                callback.writeLine('stdout', f'count: {{rows}}')
+            INPUT null
+            OUTPUT ruleExecOut
+            '''))
+
+        # Execute the rule.
+        rep_instance = IrodsConfig().default_rule_engine_plugin + '-instance'
+        self.admin.assert_icommand(['irule', '-r', rep_instance, '-F', rule_file], 'STDOUT', ['count: 2'])
+
+    @unittest.skipUnless(plugin_name == 'irods_rule_engine_plugin-python', 'Requires PREP.')
+    def test_genquery_iterator_supports_copying_genquery2_query_object__issue_7909(self):
+        rule_file = f'{self.admin.local_session_dir}/test_genquery_iterator_supports_reusing_genquery2_query_object__issue_7909.r'
+
+        with open(rule_file, 'w') as rf:
+            rf.write(dedent(f'''
+            def main(rule_args, callback, rei):
+                from genquery import Query, Parser
+                q1 = Query(callback, 'COLL_NAME', "COLL_NAME = '{self.admin.session_collection}'", parser=Parser.GENQUERY2)
+                q2 = q1.copy()
+                q3 = q2.copy()
+                iterations = 0
+                for query in [q1, q2, q3]:
+                    for r in query:
+                        callback.writeLine('stdout', f'[iteration={{iterations}}] row: {{r}}')
+                    iterations += 1
```

Comment:
If you want to get an automatic integer iterator value, you can use `enumerate` here:
```suggestion
                for iterations, query in enumerate([q1, q2, q3]):
                    for r in query:
                        callback.writeLine('stdout', f'[iteration={{iterations}}] row: {{r}}')
```

Reply:
Nice. Will update

Reply:
Done.